src/widgets/keyboard/lv_keyboard.c

```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+}
+
+#if LV_KEYBOARD_PINYIN_USE_EXT_BUTTONMATRIX
+static void lv_keyboard_free_ext_candidate_btnm_map(lv_keyboard_t * keyboard)
+{
+    if(keyboard->ext_candidate_num != 0) {
+        for(uint16_t i = 0; i < keyboard->ext_candidate_num; i++) {
+            lv_free(keyboard->ext_candidate_btnm_map[i]);
+        }
+        lv_free(keyboard->ext_candidate_btnm_map);
+        keyboard->ext_candidate_num = 0;
+    }
+}
+#endif
+
+static char * lv_keyboard_search_chinese(size_t * candidate_num, lv_keyboard_t * keyboard, const char * pinyin_str)
+{
+    /*illegal data*/
+    if(*pinyin_str == '\0')    return NULL;
+    if(*pinyin_str == 'i')     return NULL;
+    if(*pinyin_str == 'u')     return NULL;
+    if(*pinyin_str == 'v')     return NULL;
+    if(*pinyin_str == ' ')     return NULL;
+
+    size_t pinyin_str_len = strlen(pinyin_str);
+    char new_pinyin_str[pinyin_str_len + 1];
+
+    /*pinyin to lower case*/
+    for(size_t i = 0; i < pinyin_str_len; i++) {
+        if((pinyin_str[i] >= 'A') && (pinyin_str[i] <= 'Z')) {
+            new_pinyin_str[i] = pinyin_str[i] + 32;
+        }
+        else if((pinyin_str[i] >= 'a') && (pinyin_str[i] <= 'z')) {
+            new_pinyin_str[i] = pinyin_str[i];
+        }
+        else {
+            return NULL;
+        }
+    }
+
+    if(keyboard->dict == NULL) {
+        LV_LOG_ERROR("keyboard pinyin dict is NULL");
+        return NULL;
+    }
+
+    lv_keyboard_pinyin_dict_t * pinyin_head = (lv_keyboard_pinyin_dict_t *)keyboard->dict[*new_pinyin_str - 'a'];
+    lv_keyboard_pinyin_dict_t * pinyin_tail = (lv_keyboard_pinyin_dict_t *)keyboard->dict[*new_pinyin_str - 'a' + 1];
+
+    size_t i = 0;
+    for(; pinyin_head < pinyin_tail; pinyin_head++) {
+        for(i = 0; i < pinyin_str_len - 1; i++) {
+            if((*pinyin_head).pinyin_list == NULL) return NULL;
+
+            if(*((*pinyin_head).pinyin_list + i) != new_pinyin_str[i + 1]) break;
+        }
+
+        /*perfect match*/
+        if(i == (pinyin_str_len - 1)) {
+            /*The Chinese character in UTF-8 encoding format is 3 bytes*/
+            *candidate_num = strlen((const char *)(*pinyin_head).pinyin_mb_list) / 3;
+
+            return (char *)(*pinyin_head).pinyin_mb_list;
+        }
+    }
+
+    return NULL;
+}
+
+static void lv_keyboard_proc_pinyin(lv_keyboard_t * keyboard)
+{
+    char * input_pinyin = lv_label_get_text(keyboard->pinyin_label);
+    if(lv_strlen(input_pinyin) == 0) {
+        return;
+    }
+
+    size_t candidate_num = 0;
+    char * candidate_str = lv_keyboard_search_chinese(&candidate_num, keyboard, input_pinyin);
+
+    lv_keyboard_free_candidate_btnm_map(keyboard);
+    keyboard->candidate_num = candidate_num;
+
+    keyboard->candidate_btnm_map = (char **)lv_malloc((candidate_num + 1) * sizeof(char *));
+    if(keyboard->candidate_btnm_map == NULL) {
+        return;
+    }
+
+    for(size_t i = 0; i < candidate_num; i++) {
+        keyboard->candidate_btnm_map[i] = (char *)lv_malloc(4 * sizeof(char));
```

Comment:
Let make the code shorter by define a var here.

```c
char * str = lv_malloc(4);
LV_ASSERT_NULL(str);
lv_memcpy(str, candidate_str + i * 3, 3);
str[3] = '\0';
keyboard->candidate_btnm_map[i] = str;
```

Reply:
> Let make the code shorter by define a var here.
> 
> ```c
> char * str = lv_malloc(4);
> LV_ASSERT_NULL(str);
> lv_memcpy(str, candidate_str + i * 3, 3);
> str[3] = '\0';
> keyboard->candidate_btnm_map[i] = str;
> ```

okay